Source/Diagnostics/Diagnostics.cpp

```diff
@@ -39,9 +39,7 @@ Diagnostics::Diagnostics (int i, std::string name)
 {
 }
 
-Diagnostics::~Diagnostics ()
-{
-}
+Diagnostics::~Diagnostics () = default;
```

Comment:
good idea! It should be done now

Reply:
I don't exactly understand why, but the destructor of the virtual class `Diagnostics` must stay in the cpp file, otherwise I get the following error message:

```
 56%] Building CXX object CMakeFiles/WarpX.dir/Source/WarpX.cpp.o
In file included from /home/luca/Projects/WarpX/Source/WarpX.cpp:12:
In file included from /home/luca/Projects/WarpX/Source/WarpX.H:20:
In file included from /home/luca/Projects/WarpX/Source/EmbeddedBoundary/WarpXFaceInfoBox.H:11:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_Gpu.H:18:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_GpuLaunch.H:9:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_GpuError.H:5:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_Utility.H:5:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_BLassert.H:6:
In file included from /home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX.H:15:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/memory:76:
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:83:16: error: invalid application of 'sizeof' to an incomplete type 'FlushFormat'
        static_assert(sizeof(_Tp)>0,
                      ^~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of member function 'std::default_delete<FlushFormat>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/home/luca/Projects/WarpX/Source/Diagnostics/Diagnostics.H:35:13: note: in instantiation of member function 'std::unique_ptr<FlushFormat>::~unique_ptr' requested here
    virtual ~Diagnostics () = default;
            ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of member function 'std::default_delete<Diagnostics>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_construct.h:151:19: note: in instantiation of member function 'std::unique_ptr<Diagnostics>::~unique_ptr' requested here
      __pointer->~_Tp();
                  ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_construct.h:196:2: note: in instantiation of function template specialization 'std::_Destroy_aux<false>::__destroy<std::unique_ptr<Diagnostics> *>' requested here
        __destroy(__first, __last);
        ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/alloc_traits.h:848:7: note: in instantiation of function template specialization 'std::_Destroy<std::unique_ptr<Diagnostics> *>' requested here
      _Destroy(__first, __last);
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:680:7: note: in instantiation of function template specialization 'std::_Destroy<std::unique_ptr<Diagnostics> *, std::unique_ptr<Diagnostics>>' requested here
        std::_Destroy(this->_M_impl._M_start, this->_M_impl._M_finish,
             ^
/home/luca/Projects/WarpX/build/_deps/fetchedamrex-src/Src/Base/AMReX_Vector.H:25:7: note: in instantiation of member function 'std::vector<std::unique_ptr<Diagnostics>>::~vector' requested here
class Vector
      ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:361:4: note: in instantiation of member function 'std::default_delete<MultiDiagnostics>::operator()' requested here
          get_deleter()(std::move(__ptr));
          ^
/home/luca/Projects/WarpX/Source/WarpX.cpp:226:8: note: in instantiation of member function 'std::unique_ptr<MultiDiagnostics>::~unique_ptr' requested here
WarpX::WarpX ()
       ^
/home/luca/Projects/WarpX/Source/Diagnostics/FlushFormats/FlushFormat_fwd.H:11:7: note: forward declaration of 'FlushFormat'
class FlushFormat;

```
Therefore I reverted to the previous  version `Diagnostics::~Diagnostics (){}` in the cpp file...

Reply:
Hm, as long as it stays virtual it should work. But here is fine, too. Keep `= default;` though.